grpc-client-server/README.md

```diff
@@ -0,0 +1,161 @@
+
+Read the [gRPC Introduction](https://grpc.io/docs/what-is-grpc/introduction/) and 
+[gRPC Core Concepts](https://grpc.io/docs/what-is-grpc/core-concepts/) for an overview of gRPC.
+
+## Run the gRPC Quick Start Hello World Example
+
+Run through the [gRPC Hello World example](https://grpc.io/docs/languages/go/quickstart/) from the grpc.io documentation.
+
+This will ensure you have the right tools on your machine and working correctly.
+
+## Build a gRPC based prober
+
+Next, we will implement a simple prober service. Imagine that we want to verify that our site is available and has acceptable latency from many different locations around the world. We build a program that performs HTTP gets on a provided endpoint and returns statistics about how long it took.
+
+In a real production system could run several instances of our prober server in different regions and use one client to query all of the prober servers.
+
+In the same directory as this README you will find initial versions of:
+ * the protocol buffer definition
+ * prober server Go code
+ * prober client Go code
+
+Generate the generated protobuf code: 
+```console
+> protoc --go_out=. --go_opt=paths=source_relative \
+    --go-grpc_out=. --go-grpc_opt=paths=source_relative \
+    prober/prober.proto
+```
+
+Observe the new generated files:
+```console
+> ls prober
+> prober.pb.go		prober.proto		prober_grpc.pb.go
+```
+
+Read through `prober_grpc.pb.go` - this is the interface we will use in our code. This is how gRPC works: a `proto` format 
+gets generated into language-specific code that we can use to interact with gRPCs. If we're working with multiple programming languages
+that need to interact through RPCs, we can do this by generating from the same protocol buffer definition with the language-specific
+tooling. 
+
+Now run the server and client code. You should see output like this.
+
+```console
+> go run prober_server/main.go
+> 2022/10/19 17:51:32 server listening at [::]:50051
+````
+
+```console
+> go run prober_client/main.go 
+> 2022/10/19 17:52:15 Response Time: 117.000000
+```
+
+We've now gained some experience with the protocol buffer format, learned 
+how to generate Go code from protocol buffer definitions, and call that code from Go programs.
+
+## Implement prober logic
+
+Let's modify the prober service slightly. Instead of the simple one-off HTTP GET against a hardcoded google.com, we are going to modify the service to probe a HTTP endpoint N times and return the average time to GET that endpoint to the client. 
+
+Change your prober request and response to look like this and regenerate your generated go code.
+```
+// The request message
+message ProbeRequest {
+  string endpoint = 1;    // Endpoint to probe
+  int32 repetitions = 2;  // How many probes to execute
+}
+
+// The response message 
+message ProbeReply {
+  float result = 1;    // Average response time (in milliseconds)
+}
+```
```

Comment:
I'd be tempted to just instruct rather than show here:

```suggestion
Change your prober request and response:
* Add a field to the `ProbeRequest` for the number of requests to make.
* Rename the field in `ProbeReply` (and perhaps add a comment) to make clear it's the _average_ response time of the several requests.

Note that it's ok to rename fields in protobuf (unlike when we use JSON), because the binary encoding of protobuf messages doesn't include field names. You can [read more about backward/forward compatibility with protobufs](https://earthly.dev/blog/backward-and-forward-compatibility/) if you want.

Remember that you'll need to re-generate your Go code after changing your proto definitions.
```